File: services/py/quant_core/domain/pure/derive_metrics.py
"""Pure derivations from :class:`StockMeta` + latest close price.

Used by the snapshot Flight op (`list_stock_snapshots`) to produce the
list-view payload without writing derivative ratios into parquet. Every
formula is documented in ``docs/modules/01-stock-meta.md`` §2.1; this
module is the **single source of truth** for the口径 — any UI / analytics
caller must go through it instead of re-implementing.

Design rules:
    - Pure: no IO, no logging, no clock; takes inputs, returns
      :class:`DerivedMetrics`.
    - Permissive on inputs: any missing field / non-positive denominator
      → that metric is ``None``. Callers that need "x or default" wrap
      the call themselves.
    - Decimal end-to-end: never coerce through ``float`` to avoid silent
      precision loss on quantities like total_share * price.
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from quant_core.domain.types.stock import QuarterlyFinancials, StockMeta


@dataclass(frozen=True, slots=True)
class DerivedMetrics:
    """Price-derived list-view metrics. Each field is ``None`` when any
    input is missing or any denominator is ≤ 0."""

    mkt_cap: Decimal | None
    float_mkt_cap: Decimal | None
    pe_ttm: Decimal | None
    pe_dynamic: Decimal | None
    pb: Decimal | None
    peg: Decimal | None
    gross_margin_ttm: Decimal | None


_EMPTY: DerivedMetrics = DerivedMetrics(
    mkt_cap=None,
    float_mkt_cap=None,
    pe_ttm=None,
    pe_dynamic=None,
    pb=None,
    peg=None,
    gross_margin_ttm=None,
)


def derive_metrics(meta: StockMeta, price: Decimal | None) -> DerivedMetrics:
    """Compute the list-view derived metrics.

    Args:
        meta: Fully or partially populated stock meta.
        price: Latest forward-adjusted close (`close_qfq`) for the same
            ``code``, or ``None`` when the kline cache has nothing.

    Returns:
        :class:`DerivedMetrics` with every field nullable. ``None`` price
        propagates to every metric; populated price still leaves
        financial-dependent metrics ``None`` until enrichment lands.
    """
    if price is None or price <= 0:
        return _EMPTY

    mkt_cap = _mul(meta.total_share, price)
    float_mkt_cap = _mul(meta.float_share, price)

    pe_ttm = _safe_div(mkt_cap, _sum_net_profit(meta.quarterlies, -4, None))
    pe_dynamic = _pe_dynamic(meta, mkt_cap)
    pb = _safe_div(mkt_cap, meta.net_assets)
    peg = _peg(meta, pe_ttm)
    gross_margin = _gross_margin_ttm(meta.quarterlies)

    return DerivedMetrics(
        mkt_cap=mkt_cap,
        float_mkt_cap=float_mkt_cap,
        pe_ttm=pe_ttm,
        pe_dynamic=pe_dynamic,
        pb=pb,
        peg=peg,
        gross_margin_ttm=gross_margin,
    )


# -- helpers ----------------------------------------------------------------


def _mul(a: Decimal | None, b: Decimal | None) -> Decimal | None:
    if a is None or b is None:
        return None
    if a <= 0 or b <= 0:
        return None
    return a * b


def _safe_div(numerator: Decimal | None, denominator: Decimal | None) -> Decimal | None:
    if numerator is None or denominator is None:
        return None
    if denominator <= 0:
        return None
    return numerator / denominator


def _sum_net_profit(
    quarters: tuple[QuarterlyFinancials, ...], start: int, stop: int | None
) -> Decimal | None:
    """Sum ``net_profit`` over ``quarters[start:stop]``.

    Returns ``None`` when fewer than 4 quarters fall in the window or
    any quarter is missing ``net_profit``. ``slice`` was the natural
    fit but mypy --strict rejects bare ``slice`` as ``slice[Any]``;
    int boundaries keep the typing explicit.
    """
    chunk = quarters[start:stop] if stop is not None else quarters[start:]
    if len(chunk) < 4:
        return None
    total = Decimal(0)
    for q in chunk:
        if q.net_profit is None:
            return None
        total += q.net_profit
    return total


def _pe_dynamic(meta: StockMeta, mkt_cap: Decimal | None) -> Decimal | None:
    """EastMoney-style 动态 PE.

    Annualises the latest quarter's net profit using ``net_profit *
    4 / quarter_index``, where ``quarter_index ∈ {1,2,3,4}`` is derived
    from ``period.month`` (3→1, 6→2, 9→3, 12→4). The choice of口径 is
    documented in ``docs/modules/01-stock-meta.md`` §2.1 and §9; do not
    inline an alternative here.
    """
    if mkt_cap is None or not meta.quarterlies:
        return None
    latest = meta.quarterlies[-1]
    if latest.net_profit is None or latest.net_profit <= 0:
        return None
    quarter_index = _quarter_index(latest.period.month)
    if quarter_index is None:
        return None
    annualised = latest.net_profit * Decimal(4) / Decimal(quarter_index)
    if annualised <= 0:
        return None
    return mkt_cap / annualised


def _peg(meta: StockMeta, pe_ttm: Decimal | None) -> Decimal | None:
    """PEG using TTM-vs-prior-TTM growth on net_profit.

    Needs ≥ 8 reporting periods for a clean YoY comparison; fewer
    quarters / negative prior-period TTM / non-positive growth all
    propagate to ``None``.
    """
    if pe_ttm is None or len(meta.quarterlies) < 8:
        return None
    recent = _sum_net_profit(meta.quarterlies, -4, None)
    prior = _sum_net_profit(meta.quarterlies, -8, -4)
    if recent is None or prior is None or prior <= 0:
        return None
    growth_pct = (recent - prior) / prior * Decimal(100)
    if growth_pct <= 0:
        return None
    return pe_ttm / growth_pct


def _gross_margin_ttm(
    quarters: tuple[QuarterlyFinancials, ...],
) -> Decimal | None:
    if len(quarters) < 4:
        return None
    chunk = quarters[-4:]
    rev_total = Decimal(0)
    cost_total = Decimal(0)
    for q in chunk:
        if q.revenue is None or q.operating_cost is None:
            return None
        rev_total += q.revenue
        cost_total += q.operating_cost
    if rev_total <= 0:
        return None
    return (rev_total - cost_total) / rev_total


def _quarter_index(month: int) -> int | None:
    """Map a quarter-end month to its 1-indexed quarter number."""
    return {3: 1, 6: 2, 9: 3, 12: 4}.get(month)

File: apps/api/src/modules/stock-meta/domain/pure/derive-metrics.ts
/**
 * Pure port of ``services/py/quant_core/domain/pure/derive_metrics.py``.
 *
 * Both implementations are kept in sync until the snapshot Flight op
 * (`list_stock_snapshots`) also migrates; Python still uses the original
 * for its fallback path, so any口径 change must land in both files.
 * The parity check lives in ``derive-metrics.parity.test.ts`` (planned
 * Phase 2 follow-up — see ``docs/perf/storage-unify-rollout.md``).
 *
 * Permissive on inputs: any missing field / non-positive denominator
 * propagates to ``null`` so the UI's "—" path stays correct.
 */

import type { QuarterlyFinancials, StockMetaDto } from '@quant/shared';

import { D, type Dec } from './decimal-config.js';

export interface DerivedMetrics {
  readonly mkt_cap: Dec | null;
  readonly float_mkt_cap: Dec | null;
  readonly pe_ttm: Dec | null;
  readonly pe_dynamic: Dec | null;
  readonly pb: Dec | null;
  readonly peg: Dec | null;
  readonly gross_margin_ttm: Dec | null;
}

const EMPTY: DerivedMetrics = {
  mkt_cap: null,
  float_mkt_cap: null,
  pe_ttm: null,
  pe_dynamic: null,
  pb: null,
  peg: null,
  gross_margin_ttm: null,
};

export function deriveMetrics(meta: StockMetaDto, price: Dec | null): DerivedMetrics {
  if (price === null || price.lte(0)) return EMPTY;

  const totalShare = parseOpt(meta.total_share);
  const floatShare = parseOpt(meta.float_share);
  const netAssets = parseOpt(meta.net_assets);

  const mktCap = mul(totalShare, price);
  const floatMktCap = mul(floatShare, price);

  const peTtm = safeDiv(mktCap, sumNetProfit(meta.quarterlies, -4, undefined));
  const peDynamic = peDynamicCompute(meta.quarterlies, mktCap);
  const pb = safeDiv(mktCap, netAssets);
  const peg = pegCompute(meta.quarterlies, peTtm);
  const grossMargin = grossMarginTtm(meta.quarterlies);

  return {
    mkt_cap: mktCap,
    float_mkt_cap: floatMktCap,
    pe_ttm: peTtm,
    pe_dynamic: peDynamic,
    pb,
    peg,
    gross_margin_ttm: grossMargin,
  };
}

function parseOpt(value: string | null): Dec | null {
  if (value === null) return null;
  return new D(value);
}

function mul(a: Dec | null, b: Dec | null): Dec | null {
  if (a === null || b === null) return null;
  if (a.lte(0) || b.lte(0)) return null;
  return a.mul(b);
}

function safeDiv(numerator: Dec | null, denominator: Dec | null): Dec | null {
  if (numerator === null || denominator === null) return null;
  if (denominator.lte(0)) return null;
  return numerator.div(denominator);
}

/**
 * Sum ``net_profit`` over a fixed-size slice of trailing quarters.
 *
 * Returns ``null`` when fewer than 4 quarters fall in the window or
 * any quarter is missing ``net_profit`` — same口径 as Py
 * ``_sum_net_profit``.
 */
function sumNetProfit(
  quarters: readonly QuarterlyFinancials[],
  start: number,
  stop: number | undefined,
): Dec | null {
  const chunk =
    stop === undefined ? sliceFromEnd(quarters, start) : sliceFromEnds(quarters, start, stop);
  if (chunk.length < 4) return null;
  let total = new D(0);
  for (const q of chunk) {
    if (q.net_profit === null) return null;
    total = total.add(new D(q.net_profit));
  }
  return total;
}

/** Mirrors Python's ``quarters[start:]`` for negative ``start``. */
function sliceFromEnd<T>(arr: readonly T[], start: number): readonly T[] {
  return arr.slice(start < 0 ? Math.max(0, arr.length + start) : start);
}

/** Mirrors Python's ``quarters[start:stop]`` for negative bounds. */
function sliceFromEnds<T>(arr: readonly T[], start: number, stop: number): readonly T[] {
  const s = start < 0 ? Math.max(0, arr.length + start) : start;
  const e = stop < 0 ? Math.max(0, arr.length + stop) : stop;
  return arr.slice(s, e);
}

/**
 * EastMoney-style 动态 PE.
 *
 * Annualises the latest quarter's net profit using ``net_profit *
 * 4 / quarter_index``, where ``quarter_index ∈ {1,2,3,4}`` is derived
 * from ``period`` (3→1, 6→2, 9→3, 12→4). The口径 is documented in
 * ``docs/modules/01-stock-meta.md`` §2.1 and §9.
 */
function peDynamicCompute(
  quarters: readonly QuarterlyFinancials[],
  mktCap: Dec | null,
): Dec | null {
  if (mktCap === null || quarters.length === 0) return null;
  const latest = quarters[quarters.length - 1]!;
  if (latest.net_profit === null) return null;
  const netProfit = new D(latest.net_profit);
  if (netProfit.lte(0)) return null;
  const qIdx = quarterIndex(latest.period);
  if (qIdx === null) return null;
  const annualised = netProfit.mul(4).div(qIdx);
  if (annualised.lte(0)) return null;
  return mktCap.div(annualised);
}

/**
 * PEG using TTM-vs-prior-TTM growth on net_profit. Needs ≥ 8 reporting
 * periods; fewer quarters / negative prior TTM / non-positive growth
 * propagate to ``null``.
 */
function pegCompute(quarters: readonly QuarterlyFinancials[], peTtm: Dec | null): Dec | null {
  if (peTtm === null || quarters.length < 8) return null;
  const recent = sumNetProfit(quarters, -4, undefined);
  const prior = sumNetProfit(quarters, -8, -4);
  if (recent === null || prior === null || prior.lte(0)) return null;
  const growthPct = recent.sub(prior).div(prior).mul(100);
  if (growthPct.lte(0)) return null;
  return peTtm.div(growthPct);
}

function grossMarginTtm(quarters: readonly QuarterlyFinancials[]): Dec | null {
  if (quarters.length < 4) return null;
  const chunk = quarters.slice(-4);
  let revTotal = new D(0);
  let costTotal = new D(0);
  for (const q of chunk) {
    if (q.revenue === null || q.operating_cost === null) return null;
    revTotal = revTotal.add(new D(q.revenue));
    costTotal = costTotal.add(new D(q.operating_cost));
  }
  if (revTotal.lte(0)) return null;
  return revTotal.sub(costTotal).div(revTotal);
}

function quarterIndex(period: string): number | null {
  // period is ISO YYYY-MM-DD; quarter end months are 3 / 6 / 9 / 12.
  const month = Number.parseInt(period.slice(5, 7), 10);
  switch (month) {
    case 3:
      return 1;
    case 6:
      return 2;
    case 9:
      return 3;
    case 12:
      return 4;
    default:
      return null;
  }
}
